<a href="https://colab.research.google.com/github/Hion-cy/ClassFiles/blob/main/NPL_Representacion_textos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Maestría en Inteligencia Artificial y Analítica de Datos
##Procesamiento del lenguaje natural
**Matrícula:** AL263158

**Alumna:** Carmen Yolanda Hion Vela


###Instrucciones
Considera el pequeño corpus de reseñas de películas y comentarios de redes sociales que está al final del documento y realiza los siguientes ejercicios.

In [131]:
corpus_es = [
    "Me encantó la película, los actores fueron increíbles",
    "No me gustó la película, el guion fue muy malo",
    "La actuación fue excelente, pero la historia era predecible",
    "Película aburrida, me dormí a la mitad",
    "¡Maravillosa! Recomiendo esta película a todos mis amigos",
    "El guion estaba mal escrito, pero la actuación salvó la película",
    "Demasiado larga y lenta, no la volvería a ver",
    "Me gustó mucho, los efectos especiales fueron impresionantes",
    "No es mala, pero esperaba algo más emocionante",
    "Una obra maestra del cine español, increíble dirección",
    "El final fue confuso, no entendí nada",
    "Excelente cinematografía, pero el guion flojo",
    "Demasiado predecible, ya sabía lo que iba a pasar",
    "Me reí mucho, muy divertida y entretenida",
    "La música fue espectacular, pero los actores no convencieron",
    "Película mediocre, no aporta nada nuevo",
    "Excelente historia, emocionante hasta el final",
    "No me gustó la ambientación, parecía de bajo presupuesto",
    "Muy buena dirección y fotografía, la recomiendo",
    "El guion tenía agujeros, pero los efectos visuales impresionan",
    "Aburrida, diálogo poco natural",
    "Gran actuación de los protagonistas, realmente me impactó",
    "El ritmo es lento, pero la historia es interesante",
    "No la recomiendo, desperdicié mi tiempo",
    "Un clásico moderno, me encantó cada escena",
    "Las escenas de acción fueron espectaculares",
    "El humor es pobre y los personajes poco creíbles",
    "Me emocioné, la historia me llegó al corazón",
    "La trama es confusa y difícil de seguir",
    "Película fantástica, muy bien lograda"
]

###Ejercicio 1

#### Transformación
* Crear un CountVectorizer y un TfidfVectorizer.
* Aplícalos al corpus y transformar los textos.

In [132]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
import pandas as pd

In [142]:
vectorizer= CountVectorizer()
X=vectorizer.fit_transform(corpus_es)


bow_df= pd.DataFrame(
    X.toarray(),
    columns=vectorizer.get_feature_names_out(),
    index=[f"d{i+1}" for i in range(len(corpus_es))]
)

#print(X)
#print(X.toarray(),"\n")
#print(list(bow_df.columns))
#print(vectorizer.get_feature_names_out())
print("---- CountVectorizer ----")
print(bow_df,"\n")

---- CountVectorizer ----
     aburrida  acción  actores  actuación  agujeros  al  algo  ambientación  \
d1          0       0        1          0         0   0     0             0   
d2          0       0        0          0         0   0     0             0   
d3          0       0        0          1         0   0     0             0   
d4          1       0        0          0         0   0     0             0   
d5          0       0        0          0         0   0     0             0   
d6          0       0        0          1         0   0     0             0   
d7          0       0        0          0         0   0     0             0   
d8          0       0        0          0         0   0     0             0   
d9          0       0        0          0         0   0     1             0   
d10         0       0        0          0         0   0     0             0   
d11         0       0        0          0         0   0     0             0   
d12         0       0     

In [134]:
vectorizer = TfidfVectorizer()
X = vectorizer.fit_transform(corpus_es) # Convertir a DataFrame
Tfid_df = pd.DataFrame(
     X.toarray(),
     columns=vectorizer.get_feature_names_out() )
print("---- TfidfVectorizer ----")
print(Tfid_df, "\n")

---- TfidfVectorizer ----
    aburrida    acción   actores  actuación  agujeros        al      algo  \
0   0.000000  0.000000  0.413604   0.000000    0.0000  0.000000  0.000000   
1   0.000000  0.000000  0.000000   0.000000    0.0000  0.000000  0.000000   
2   0.000000  0.000000  0.000000   0.344901    0.0000  0.000000  0.000000   
3   0.460672  0.000000  0.000000   0.000000    0.0000  0.000000  0.000000   
4   0.000000  0.000000  0.000000   0.000000    0.0000  0.000000  0.000000   
5   0.000000  0.000000  0.000000   0.301442    0.0000  0.000000  0.000000   
6   0.000000  0.000000  0.000000   0.000000    0.0000  0.000000  0.000000   
7   0.000000  0.000000  0.000000   0.000000    0.0000  0.000000  0.000000   
8   0.000000  0.000000  0.000000   0.000000    0.0000  0.000000  0.404267   
9   0.000000  0.000000  0.000000   0.000000    0.0000  0.000000  0.000000   
10  0.000000  0.000000  0.000000   0.000000    0.0000  0.000000  0.000000   
11  0.000000  0.000000  0.000000   0.000000    0.0

####Mostrar

1.	Dimensiones de la matriz resultante.
2.	Primeros 10 tokens y sus índices.


In [135]:
print(f"Dimensiones BoW (Filas, Columnas): {bow_df.shape}")
print(f"Dimensiones TF-IDF (Filas, Columnas): {Tfid_df.shape}")

Dimensiones BoW (Filas, Columnas): (30, 127)
Dimensiones TF-IDF (Filas, Columnas): (30, 127)


In [136]:
tokens = vectorizer.get_feature_names_out()
print("---- Primeros 10 tokens y sus índices ----")
for i in range(10):
    print(f"Índice {i}: {tokens[i]}")


---- Primeros 10 tokens y sus índices ----
Índice 0: aburrida
Índice 1: acción
Índice 2: actores
Índice 3: actuación
Índice 4: agujeros
Índice 5: al
Índice 6: algo
Índice 7: ambientación
Índice 8: amigos
Índice 9: aporta


####Comparar las diferencias entre ambas matrices (CountVectorizer y TfidfVectorizer), mencionando de manera general las diferencias observadas.

**Conclusión:** Al no integrar ningún parámetro en la realizacion de CountVectorizer y TfidfVectorizer el tamaño de ambas matrices se mantiene igual para ambas transformaciones (30x127).

Sin embargo existen palabras como "aburrida" donde, si solo tomáramos en cuenta el primero (Count) no sería muy útil ya que en casos como d4 y d21 ambos poseen el valor de 1, mientras que en el segundo (Tfid) el peso varía entre 0.460672 y 0.470578 donde se puede afirmar que la palabra tiene más peso en la segunda reseña (d21 o 20).

Otros ejemplos como actuación (d6 (5) y d22 (21)) refuerzan la diferencia entre emplear un método u otro para valorar la importancia de cada palabra.


### Ejercicio 2

Preprocesa el texto eliminando los acentos y stopwords, así como convirtiéndolo a minúsculas.

In [137]:
import nltk
from nltk.corpus import stopwords

nltk.download('stopwords')
stop_words_es = stopwords.words('spanish')


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


#### Realiza las mismas actividades que en el ejercicio anterior.

In [138]:

vectorizer_parametros= CountVectorizer(lowercase=True, stop_words=stop_words_es, strip_accents='unicode' )
X_parametros=vectorizer_parametros.fit_transform(corpus_es)

print("---- CountVectorizer ----")
bow_param= pd.DataFrame(
    X_parametros.toarray(),
    columns=vectorizer_parametros.get_feature_names_out(),
    index=[f"d{i+1}" for i in range(len(corpus_es))]
)

print(bow_param,"\n")


---- CountVectorizer ----
     aburrida  accion  actores  actuacion  agujeros  ambientacion  amigos  \
d1          0       0        1          0         0             0       0   
d2          0       0        0          0         0             0       0   
d3          0       0        0          1         0             0       0   
d4          1       0        0          0         0             0       0   
d5          0       0        0          0         0             0       1   
d6          0       0        0          1         0             0       0   
d7          0       0        0          0         0             0       0   
d8          0       0        0          0         0             0       0   
d9          0       0        0          0         0             0       0   
d10         0       0        0          0         0             0       0   
d11         0       0        0          0         0             0       0   
d12         0       0        0          0         

/usr/local/lib/python3.12/dist-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['eramos', 'estabamos', 'estais', 'estan', 'estara', 'estaran', 'estaras', 'estare', 'estareis', 'estaria', 'estariais', 'estariamos', 'estarian', 'estarias', 'esteis', 'esten', 'estes', 'estuvieramos', 'estuviesemos', 'fueramos', 'fuesemos', 'habeis', 'habia', 'habiais', 'habiamos', 'habian', 'habias', 'habra', 'habran', 'habras', 'habre', 'habreis', 'habria', 'habriais', 'habriamos', 'habrian', 'habrias', 'hayais', 'hubieramos', 'hubiesemos', 'mas', 'mia', 'mias', 'mio', 'mios', 'seais', 'sera', 'seran', 'seras', 'sere', 'sereis', 'seria', 'seriais', 'seriamos', 'serian', 'serias', 'si', 'tambien', 'tendra', 'tendran', 'tendras', 'tendre', 'tendreis', 'tendria', 'tendriais', 'tendriamos', 'tendrian', 'tendrias', 'teneis', 'tengais', 'tenia', 'teniais', 'teniamos', 'tenian', 'tenias', 'tuvie

In [139]:

vectorizer_parametros= TfidfVectorizer(lowercase=True,  stop_words=stop_words_es, strip_accents='unicode' )
X_parametros=vectorizer_parametros.fit_transform(corpus_es)

Tfid_param = pd.DataFrame(
     X_parametros.toarray(),
     columns=vectorizer_parametros.get_feature_names_out() )
print("---- TfidfVectorizer ----")
print(Tfid_param, "\n")

---- TfidfVectorizer ----
    aburrida   accion   actores  actuacion  agujeros  ambientacion   amigos  \
0   0.000000  0.00000  0.515968   0.000000   0.00000      0.000000  0.00000   
1   0.000000  0.00000  0.000000   0.000000   0.00000      0.000000  0.00000   
2   0.000000  0.00000  0.000000   0.496498   0.00000      0.000000  0.00000   
3   0.499118  0.00000  0.000000   0.000000   0.00000      0.000000  0.00000   
4   0.000000  0.00000  0.000000   0.000000   0.00000      0.000000  0.57167   
5   0.000000  0.00000  0.000000   0.378626   0.00000      0.000000  0.00000   
6   0.000000  0.00000  0.000000   0.000000   0.00000      0.000000  0.00000   
7   0.000000  0.00000  0.000000   0.000000   0.00000      0.000000  0.00000   
8   0.000000  0.00000  0.000000   0.000000   0.00000      0.000000  0.00000   
9   0.000000  0.00000  0.000000   0.000000   0.00000      0.000000  0.00000   
10  0.000000  0.00000  0.000000   0.000000   0.00000      0.000000  0.00000   
11  0.000000  0.00000  0.0

/usr/local/lib/python3.12/dist-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['eramos', 'estabamos', 'estais', 'estan', 'estara', 'estaran', 'estaras', 'estare', 'estareis', 'estaria', 'estariais', 'estariamos', 'estarian', 'estarias', 'esteis', 'esten', 'estes', 'estuvieramos', 'estuviesemos', 'fueramos', 'fuesemos', 'habeis', 'habia', 'habiais', 'habiamos', 'habian', 'habias', 'habra', 'habran', 'habras', 'habre', 'habreis', 'habria', 'habriais', 'habriamos', 'habrian', 'habrias', 'hayais', 'hubieramos', 'hubiesemos', 'mas', 'mia', 'mias', 'mio', 'mios', 'seais', 'sera', 'seran', 'seras', 'sere', 'sereis', 'seria', 'seriais', 'seriamos', 'serian', 'serias', 'si', 'tambien', 'tendra', 'tendran', 'tendras', 'tendre', 'tendreis', 'tendria', 'tendriais', 'tendriamos', 'tendrian', 'tendrias', 'teneis', 'tengais', 'tenia', 'teniais', 'teniamos', 'tenian', 'tenias', 'tuvie

In [140]:
print(f"Dimensiones BoW (Filas, Columnas): {bow_param.shape}")
print(f"Dimensiones TF-IDF (Filas, Columnas): {Tfid_param.shape}")

Dimensiones BoW (Filas, Columnas): (30, 97)
Dimensiones TF-IDF (Filas, Columnas): (30, 97)


In [141]:
tokens_param = vectorizer_parametros.get_feature_names_out()
print("---- Primeros 10 tokens y sus índices ----")
for i in range(10):
    print(f"Índice original {i+1}: {tokens[i]}")
    print(f"Índice {i+1}: {tokens_param[i]}\n")


---- Primeros 10 tokens y sus índices ----
Índice original 1: aburrida
Índice 1: aburrida

Índice original 2: acción
Índice 2: accion

Índice original 3: actores
Índice 3: actores

Índice original 4: actuación
Índice 4: actuacion

Índice original 5: agujeros
Índice 5: agujeros

Índice original 6: al
Índice 6: ambientacion

Índice original 7: algo
Índice 7: amigos

Índice original 8: ambientación
Índice 8: aporta

Índice original 9: amigos
Índice 9: bajo

Índice original 10: aporta
Índice 10: bien



####	Compara las matrices del ejercicio anterior con las matrices que obtuviste en este ejercicio, mencionando de manera general las diferencias observadas.

**Conclusión:** La diferencia principal entre las matrices originales y las matrices "limpias" es que al eliminar las stop-words el tamaño se reduce de 30x127 a 30x97.

Además al eliminar los acentos se unifican palabras como "acción" y "acción" para interpretarse como la misma palabra permitiendo que el conteo y peso se repartan de manera más uniforme.

Los índices también cambian, en el caso del índice 6 que originalmente era "al" paso a ser "ambientación" (en el original ocupa el lugar 8) que da un contexto más acercado al tema (películas). Lo mismo ocurre con "amigos" que paso del lugar 9 al 7.
